### PR TITLE
Create PHP release workflow in drone

### PR DESCRIPTION
**Pull request by nupplaphil**
As already discussed with @tobiasd , this would be a solution I found instead of the github workflow.

This pipeline would get triggered by a cron-job (see https://docs.drone.io/cron/ ) and not with every PR.
I already created a cron for `develop`, which would run every <strike>15 minutes</strike> night.

It includes the following steps:
- `Check changes` ... checks if something changed at the current git branch for the last 15 minutes
- `Create artifacts` ... creates a tar.gz with just releasable content (see `mods/release-list.txt` for the whitelist) and a SHA256 checksum file
- `Upload artifacts` ... Would upload the two artifacts and uses a private key wit a private key password, based on encrypted secrets on the drone server

@utzer / @tobiasd  - Is the `files.friendi.ca` on the same server as the `drone.friendi.ca` VM ? If so, we could create a specific user for drone which is just allowed for a specific folder and is only accessible from the internal IP of the `drone.friendi.ca` VM

### CODE REVIEW

**Comment by nupplaphil** [comment]
I talked with @utzer about this PR.
The 2 domains are different VMs on the same root-server

There are 2 possible solutions
- creating an SSH user and add the restrictions I described (the difficult part is to allow the access to a www-data owned directory)
- moving the `files.friendi.ca` on an own docker container at the same VM as `drone.friendi.ca`, so there's no need for SSH and just use a shared volume for drone and files

the second solution is less complex, but because I'm just starting with the develop-artifact generation, we do need @tobiasd scripts for the official release-artifacts & addon-releases yet. Maybe this solution is applicable when all artifacts are automatically created because of GIT events

**Comment by MrPetovan** [comment]
Do we need 15-minutes frequency for develop? Daily builds are enough for this purpose, since develop can be pulled from Git anyway. Stable builds cron can be that frequent since there isn't much action on the stable branch front, unless the build takes close to 15 minutes.

**Comment by tobiasd** [comment]
I'll have a look to create a user with the sufficient rights on the server to put the files into a directory but not much more. The files for the development/RC branch will be stored into a separate subdirectory of the files, or do you want to mix them in one directory?

For my taste it would be enough to have a nightly build. Fully aware of the consequence that then a broken day will then last at least a day even if the repository might already include a fix for the problem.

**Comment by nupplaphil** [comment]
I thought 15 minutes would be good to have most recent archives fast in case of troubles.
But a "nightly build", as I know it's standard for developments, are fine for me too :-) And If needed, we could execute the job on demand as well, this gives us the flexibility in case of troubles

I'd say there's no need to have different folders. There's "just" one nightly build, named `friendica-full-develop.tar.gz`or `friendica-full-2021.06-rc.tar.gz`. Otherwise we'd have to create some "cleanup jobs" as well.

@tobiasd / @utzer - Did you solve the user permission problem yet? I know you can "chroot" users and link the commands into the chrooted directory for SSH. I found a possible solution here: https://www.thegeekstuff.com/2012/03/chroot-sftp-setup/

**Comment by tobiasd** [comment]
On Fri, 28 May 2021 10:17:07 -0700
Philipp ***@***.***> wrote:

> @tobiasd / @utzer - Did you solve the user permission problem yet? I
> know you can "chroot" users and link the commands into the chrooted
> directory for SSH. I found a possible solution here:
> https://www.thegeekstuff.com/2012/03/chroot-sftp-setup/

I had no time so far to look into the user settings on the drone server - sorry! will do ASAP

**Comment by nupplaphil** [comment]
@tobiasd no worries, we're part of a hobby project, so take your time :-)

**Comment by MrPetovan** [comment]
> I thought 15 minutes would be good to have most recent archives fast in case of troubles.

What kind of troubles are you expecting that a 15-minute old develop build could solve?

**Comment by nupplaphil** [comment]
Hmm.. Afaik we sometimes brick functionallity (like Federation processing) at develop and I thought sometimes it had to be fixed quick because of the impact on large nodes..

But as said, it was just a educated guess and I'm fine with a nightly build :)

**Comment by MrPetovan** [comment]
I'm still wondering who would use the develop build archives to run a node when git is so much more convenient?

**Comment by nupplaphil** [comment]
hopefully all docker users for example *gg*
- Download full-archive
- sha256 check
- sync into `/var/html/www/`
- start

instead of 

- git checkout
- git addons checkout
- git composer install --no-dev
- sync to `/var/www/html`
- start

I think this would reduce the downtime and is more "bulletproof"
And an admin doesn't need GIT knowledge anymore to start a friendica node

**Comment by MrPetovan** [comment]
Again, we're talking about the `develop` branch. I'm happy these archives are built, but would-be admins already didn't need git knowledge to start a Friendica node. And I don't think it is far-fetched to expect admins wanting to run a `develop` node to know how to use basic git commands. Even @AlfredSK manages to do it!

**Comment by nupplaphil** [comment]
I think, a long-time Friendica experienced admin like @alfredsk isn't an good example ;-)

I think about some hobby-admins with Raspberrys f.e. and they want to "try" Friendica in the most easy way they can. Fediverse still has the slogan that new users should run their own server rather than use other servers, so every step to make it more comfortable is a good step - even for `develop`. For me, it's the message "it's the nightly build, beware of some unstable features, but you can use it the same way as a RC or the stable archives, so just try it if you like..." and they don't have to consider installing GIT on their server and clone repositories ..

**Comment by annando** [comment]
Concerning Drone: I just read this post: https://libranet.de/display/0b6b25a8-8360-b314-1553-c41073413876

**Comment by AlfredSK** [comment]
I can think of a situation where archive files of the develop branch could become handy.

An admin has installed Friendica with the archive files method. Then it turns out that there is a bug in the released stable version. In this situation the admin could simply update the node to the latest develop version using the provided archive files.

**Comment by MrPetovan** [comment]
Fair enough, even in this case it seems to me that the nightly builds are enough, or would a bug be so pressing that we don't issue a hotfix for stable and that someone wouldn't want to wait for the next nightly build?

**Comment by AlfredSK** [comment]
A nightly build is absolutely sufficient.

**Comment by nupplaphil** [comment]
> Concerning Drone: I just read this post: https://libranet.de/display/0b6b25a8-8360-b314-1553-c41073413876

örks .. I'd say I still finish the last github actions to drone migrations and ttan we could consider migrating to https://github.com/woodpecker-ci/woodpecker .. I read the whole blogpost as well: https://laszlo.cloud/drone-community-edition-what-is-included

**Comment by nupplaphil** [comment]
Switched to 2021.09 because of the upcoming release :)

**Comment by nupplaphil** [comment]
switching to 2021.12 - maybe  @tobiasd  &  @utzer and I will find a solution for the next release :-)

**Comment by nupplaphil** [comment]
Aaaand .. done :-)
--> this Workflow already created https://files.friendi.ca/friendica-full-2021.09-dev-SNAPSHOT.tar.gz :-)

**Comment by nupplaphil** [comment]
btw. after this merge, we got rid of any github dependency so far ...

**Comment by tobiasd** [comment]
The generated SNAPSHOT tar archive has the files directly in the root of the archive. For the releases we add a subdirectory for the files. Not sure if it was you who asked for that subdirectory or the yunohost packagers.

**Comment by nupplaphil** [comment]
@tobiasd  plz have a look here:
https://files.friendi.ca/friendica-full-2021.09-rc.tar.gz
- addon is now included
- subdir is included

File: .drone.yml [hunk 1]
```diff
@@ -325,3 +325,182 @@ volumes:
...
+      - apt-get update
+      - apt-get install bzip2
+      - export VERSION="$(cat VERSION)"
+      - export RELEASECORE="friendica-full-$VERSION"
+      - export ARTIFACT="$RELEASECORE.tar.gz"
+      - # Create artifact
+      - mkdir ./build
+      - tar -cvjf ./build/$ARTIFACT --transform "s,^,$RELEASECORE/," -X mods/release-list-exclude.txt -T mods/release-list-include.txt
+      - # calculate SHA256 checksum
+      - sha256sum ./build/$ARTIFACT > ./build/$ARTIFACT.sha256
```
Comment (tobiasd) [inline .drone.yml]: For me the additional `build` directory in the patch is somehow disturbing.

```
$> cat friendica-full-2021.09-rc.tar.gz.sha256; sha256sum friendica-full-2021.09-rc.tar.gz
df77fc21c6ccc224e6f548a137e0289a87dacd945fb0d3600f049570fdcc22bf  ./build/friendica-full-2021.09-rc.tar.gz
df77fc21c6ccc224e6f548a137e0289a87dacd945fb0d3600f049570fdcc22bf  friendica-full-2021.09-rc.tar.gz
```

and the same for the addon archive. Maybe the sum should be generated in the build directory.

**Comment by tobiasd** [comment]
Why include the addon repository into the archive? I like them separated more. For packaging each of the addons most likely should be in their own package with "addon" being a meta package pulling in all the single-addon packages as requirement.

**Comment by nupplaphil** [comment]
what about now:
- https://files.friendi.ca/friendica-2021.09-rc.tar.gz <-- core without addons
- https://files.friendi.ca/friendica-full-2021.09-rc.tar.gz <-- full with addons

@tobiasd and I added a sha256 output as well (see last lines of https://drone.friendi.ca/friendica/friendica/2428/9/5 ;-) )

**Comment by tobiasd** [comment]
I'll have a look, but why re-purpose the full archive to include not only the vendor folder but also the addons? If we want to introduce that package with the hole bundle I think it should bear a new name.

**Comment by nupplaphil** [comment]
Ah I see, It's good for me as well just having them separated, so I reverted the creation of the full package including `addon/` again.
--> Now the content of the "full" packages are equal to the other "full" packages so far :)

**Comment by tobiasd** [comment]
The `cat` / `sha256sum` output now looks good for me :)

File: .drone.yml [hunk 2]
```diff
@@ -325,3 +325,200 @@ volumes:
   - name: cache
     host:
       path: /tmp/drone-cache
+---
+kind: pipeline
+type: docker
+name: continous-deployment
```
Comment (MrPetovan) [inline .drone.yml]:
```suggestion
name: continuous-deployment
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**